common/src/main/java/org/conscrypt/ConscryptEngineSocket.java

```diff
@@ -615,6 +635,14 @@ public void close() throws IOException {
             ConscryptEngineSocket.this.close();
         }
 
+        void release() {
+            synchronized (readLock) {
+                if (allocatedBuffer != null) {
+                    allocatedBuffer.release();
```

Comment:
May be worthwhile to implement a free function release direct memory without waiting on the finalizer queue for the unpooled case, for example:
https://github.com/undertow-io/undertow/blob/ff4c9cf37872cb96070ba6a2fcbbaa6df291e390/core/src/main/java/io/undertow/server/DirectByteBufferDeallocator.java

Reply:
We may want to set `allocatedBuffer = null` after any time that we invoke release

Reply:
@nmittler I'm going to leave it this way just because it mirrors ConscryptEngine's usage pattern, but that's a possible change in the future.  I think it works out just fine either way, though.

@cakofony `sun.misc.Unsafe` isn't part of the public Android API, so it's not very easy to do, and we need to run on both platforms.  We also can't null out the reference because it's final (and I think the benefits of being final outweigh being able to null out the reference).